src/rustc.rs

```diff
@@ -19,20 +20,212 @@ impl TargetList {
+        Self {
+            channel: channel.to_string(),
+            date: date.clone(),
+            host: host.clone(),
+            is_custom: false,
+            full: if let Some(date) = date {
+                format!("{}-{}-{}", channel, date, host)
+            } else {
+                format!("{}-{}", channel, host)
+            },
+        }
+    }
+
+    /// Replace the host, does nothing if ran on a custom toolchain
+    pub fn replace_host(&mut self, host: &TargetTriple) -> &mut Self {
+        if !self.is_custom {
+            *self = Self::new(&self.channel, &self.date, host);
+        }
+        self
+    }
+
+    /// Makes a good guess as to what the toolchain is compiled to run on.
+    pub(crate) fn custom(name: &str, sysroot: &Path, verbose: bool) -> Result<QualifiedToolchain> {
+        if let Ok(compat) = std::env::var("CROSS_CUSTOM_TOOLCHAIN_COMPAT") {
+            let mut toolchain: QualifiedToolchain = compat.parse().wrap_err(
+                "could not parse CROSS_CUSTOM_TOOLCHAIN_COMPAT as a fully qualified toolchain name",
+            )?;
+            toolchain.is_custom = true;
+            toolchain.full = name.to_string();
+            return Ok(toolchain);
+        }
+        // a toolchain installed by https://github.com/rust-lang/cargo-bisect-rustc
+        if name.starts_with("bisector-nightly") {
+            let (_, toolchain) = name.split_once('-').expect("should include -");
+            let mut toolchain: QualifiedToolchain = toolchain
+                .parse()
+                .wrap_err("could not parse bisector toolchain")?;
+            toolchain.is_custom = true;
+            toolchain.full = name.to_string();
+            return Ok(toolchain);
+        } else if let Ok(stdout) = Command::new(sysroot.join("bin/rustc"))
+            .arg("-Vv")
+            .run_and_get_stdout(verbose)
+        {
+            let rustc_version::VersionMeta {
+                build_date,
+                channel,
+                host,
+                ..
+            } = rustc_version::version_meta_for(&stdout)?;
+            let mut toolchain = QualifiedToolchain::new(
+                match channel {
+                    rustc_version::Channel::Dev => "dev",
+                    rustc_version::Channel::Nightly => "nightly",
+                    rustc_version::Channel::Beta => "beta",
+                    rustc_version::Channel::Stable => "stable",
+                },
+                &build_date,
+                &host.into(),
+            );
+            toolchain.is_custom = true;
+            toolchain.full = name.to_string();
+            return Ok(toolchain);
+        }
+        return Err(eyre::eyre!(
+            "cross can not figure out what your custom toolchain is"
+        ))
+        .suggestion("set `CROSS_CUSTOM_TOOLCHAIN_COMPAT` to a fully qualified toolchain name: i.e `nightly-aarch64-unknown-linux-musl`");
+    }
```

Comment:
I'm not sure if we need more than this for checking what the toolchains host is. This works in most cases that are sane.

Could maybe add a check for `.*-<host_triple>` in toolchain name and then assume the channel is "nightly"

there should be atleast two folders in `sysroot/lib/rustlib`: `etc` and `<host_triple>`, if there are more, they are targets that this toolchain can compile to. one possible check is to see if there's three folders, filter out `etc` and the known specified `<target_triple>` and then you have the host. Seems a bit too much though